[JIT] Fix toIValue handling of AttributeError when casting ClassType

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49188 [JIT] Fix toIValue handling of AttributeError when casting ClassType**

Previously `toIValue` would try to pull out all the attributes from a Python object to copy them over into an IValue Class instance. However, the logic to do so relied on `py::getattr`, which would throw a Python `AttributeError` if the attribute does not exist. This is problematic since in schema matching we expect pretty specific exception types to be caught (namely `py::cast_error` from `toIValue` and `schema_match_error` on schema match failure), but this `AttributeError` was not include in that set. This PR makes it so that we check for the presence of a given attribute on the rhs Python object before calling `py::getattr`. If it does not exist, throw a `py::cast_error` to allow schema resolution to continue.

Internal: fixes https://fb.workplace.com/groups/1144215345733672/permalink/1843786182443248/

Differential Revision: [D25476573](https://our.internmc.facebook.com/intern/diff/D25476573)

Commit:
[JIT] Fix toIValue handling of AttributeError when casting ClassType

[ghstack-poisoned]

SplitInfinity reviewed (commented):
Very nice, can you add a unit test to `test/jit/test_class_type.py` for this as well?

Commit:
Update on "[JIT] Fix toIValue handling of AttributeError when casting ClassType"


Previously `toIValue` would try to pull out all the attributes from a Python object to copy them over into an IValue Class instance. However, the logic to do so relied on `py::getattr`, which would throw a Python `AttributeError` if the attribute does not exist. This is problematic since in schema matching we expect pretty specific exception types to be caught (namely `py::cast_error` from `toIValue` and `schema_match_error` on schema match failure), but this `AttributeError` was not include in that set. This PR makes it so that we check for the presence of a given attribute on the rhs Python object before calling `py::getattr`. If it does not exist, throw a `py::cast_error` to allow schema resolution to continue.

Internal: fixes https://fb.workplace.com/groups/1144215345733672/permalink/1843786182443248/

Differential Revision: [D25476573](https://our.internmc.facebook.com/intern/diff/D25476573)

[ghstack-poisoned]

Commit:
Update on "[JIT] Fix toIValue handling of AttributeError when casting ClassType"


Previously `toIValue` would try to pull out all the attributes from a Python object to copy them over into an IValue Class instance. However, the logic to do so relied on `py::getattr`, which would throw a Python `AttributeError` if the attribute does not exist. This is problematic since in schema matching we expect pretty specific exception types to be caught (namely `py::cast_error` from `toIValue` and `schema_match_error` on schema match failure), but this `AttributeError` was not include in that set. This PR makes it so that we check for the presence of a given attribute on the rhs Python object before calling `py::getattr`. If it does not exist, throw a `py::cast_error` to allow schema resolution to continue.

Internal: fixes https://fb.workplace.com/groups/1144215345733672/permalink/1843786182443248/

Differential Revision: [D25476573](https://our.internmc.facebook.com/intern/diff/D25476573)

[ghstack-poisoned]

SplitInfinity commented on test/jit/test_class_type.py:
```diff
@@ -959,6 +959,26 @@ def forward(self, x):
+
+        test_fn(Foo(3, 4.0))
+
+        with self.assertRaisesRegex(RuntimeError, 'missing attribute i'):
```
For my own information: before this change, an `AttributeError` would have been thrown here, right?

jamesr66a replied:
Right

Commit:
Update on "[JIT] Fix toIValue handling of AttributeError when casting ClassType"


Previously `toIValue` would try to pull out all the attributes from a Python object to copy them over into an IValue Class instance. However, the logic to do so relied on `py::getattr`, which would throw a Python `AttributeError` if the attribute does not exist. This is problematic since in schema matching we expect pretty specific exception types to be caught (namely `py::cast_error` from `toIValue` and `schema_match_error` on schema match failure), but this `AttributeError` was not include in that set. This PR makes it so that we check for the presence of a given attribute on the rhs Python object before calling `py::getattr`. If it does not exist, throw a `py::cast_error` to allow schema resolution to continue.

Internal: fixes https://fb.workplace.com/groups/1144215345733672/permalink/1843786182443248/

Differential Revision: [D25476573](https://our.internmc.facebook.com/intern/diff/D25476573)

[ghstack-poisoned]
